keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,437 @@
+
+[Horizontal Pod Autoscaler]: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/
+
+## Motivation
+
+Different applications may have different business values, different logic and may require different scaling behaviors.
+I can name at least three types of applications:
+
+- Applications that handle business-critical web traffic. They should scale up as fast as possible (false positive signals to scale up are ok), and scale down very slowly (waiting for another traffic spike).
+- Applications that process critical data. They should scale up as fast as possible (to reduce the data processing time), and scale down as soon as possible (to reduce the costs). False positives signals to scale up/down are ok.
+- Applications that process regular data/web traffic. It is not that important and may scale up and down in a usual way to minimize jitter.
+
+At the moment, there’s only one cluster-level configuration parameter that influence how fast the cluster is scaled down:
+
+- [--horizontal-pod-autoscaler-downscale-stabilization-window][]   (default to 5 min)
+
+And a couple of hard-coded constants that specify how fast the cluster can scale up:
+
+- [scaleUpLimitFactor][] = 2.0
+- [scaleUpLimitMinimum][] = 4.0
+
+As a result, users can't influence scale velocity, and that is a problem for many people. There're several open issues in tracker about that:
+
+- [#39090][]
+- [#65097][]
+- [#69428][]
+
+[--horizontal-pod-autoscaler-downscale-stabilization-window]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+[scaleUpLimitFactor]: https://github.com/kubernetes/kubernetes/blob/v1.13.4/pkg/controller/podautoscaler/horizontal.go#L55
+[scaleUpLimitMinimum]: https://github.com/kubernetes/kubernetes/blob/v1.13.4/pkg/controller/podautoscaler/horizontal.go#L56
+[#39090]: https://github.com/kubernetes/kubernetes/issues/39090
+[#65097]: https://github.com/kubernetes/kubernetes/issues/65097
+[#69428]: https://github.com/kubernetes/kubernetes/issues/69428
+
+### Goals
+
+- Allow the user to be able to manage the scale velocity
+
+### Non-Goals
+
+TBA
+
+## Proposal
+
+We need to introduce an algorithm-agnostic HPA object configuration that will configure each particular HPA scaling behavior.
+For both direction (scale up and scale down) there should be a `Constraint` field with the following parameters:
+
+- `periodSeconds` - a parameter to specify the time period for the constraint, in seconds
+- `percent` - a parameter to specify the relative speed, in percentages:
+    i.e., if ScaleUpPercent = 150 , then we can add 150% more pods (10 -> 25 pods)
+- `pods` - a parameter to specify the absolute speed, in the number of pods:
+    i.e., if ScaleUpPods = 5 , then we can add 5 more pods (10 -> 15)
+- `delay` - a parameter to specify a delay between the first signal to change the number of pods and the real scaling. It works very similar to the [Stabilization Window][] approach.
```

Comment:
`delay` behaves *exactly* like the stabilization window.  I think that's important and correct.

This sounds like it's similar but not the same.  How about: "a parameter to specify the window over which the max (or min) recommendation is used. It behaves the same as the Stabilization Window approach"

Reply:
Yep, thank you for this wording. Used your variant: https://github.com/kubernetes/enhancements/pull/883/commits/e47c961cbac46c4550830a7ea0f10a9c007641a9